From 4b767fa89cd5d5425dfb4a9409790ca64b9e65e4 Mon Sep 17 00:00:00 2001
From: "Bhakte, GurudattaX" <gurudattax.bhakte@intel.com>
Date: Thu, 9 Apr 2015 20:15:19 +0530
Subject: [PATCH] Vibra: [CHT]: LPSS1_F2_PWM2 remains in D0 and blocks PM

This patch disables PWM2 for CHT HR platform based on PMIC ID,
to unblock PM (s0i1, S0i3 and S3).

Change-Id: Ic8a68e61f3d786299653989b97fdbefad61b18d3
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-7709
Signed-off-by: Bhakte, GurudattaX <gurudattax.bhakte@intel.com>
---
 drivers/input/misc/intel_mid_vibra_acpi.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/input/misc/intel_mid_vibra_acpi.c b/drivers/input/misc/intel_mid_vibra_acpi.c
index 66f2f38ad6b5..63dcd1d4d90f 100644
--- a/drivers/input/misc/intel_mid_vibra_acpi.c
+++ b/drivers/input/misc/intel_mid_vibra_acpi.c
@@ -43,6 +43,8 @@
 
 #define CRYSTALCOVE_PMIC_VIBRA_MAX_BASEUNIT	0x7F
 
+#define CRYSTALCOVE_PMIC_ID_ADDR 		0x6E00
+#define CRYSTALCOVE_PMIC 			0x1F
 /* for CHT CR SOC controlled vibra */
 #define PERIOD_NS   40000
 #define DUTY_NS_ON  20000     /* 50 % */
@@ -161,6 +163,7 @@ int intel_mid_plat_vibra_probe(struct platform_device *pdev)
 	struct gpio_desc *gpio_en;
 	int ret;
 	const char *board_name;
+	u8 pmic_id;
 
 	ret = acpi_bus_get_device(handle, &device);
 	if (ret) {
@@ -176,6 +179,17 @@ int intel_mid_plat_vibra_probe(struct platform_device *pdev)
 		return -ENODEV;
 	}
 	board_name = dmi_get_system_info(DMI_BOARD_NAME);
+	if (strncmp(board_name, "Cherry Trail Tablet", DMI_STRING_MAX) == 0) {
+		pmic_id = intel_soc_pmic_readb(CRYSTALCOVE_PMIC_ID_ADDR);
+		if (pmic_id < 0)
+			pr_err("Error reading PMIC ID register\n");
+		else
+			pr_debug("PMIC-ID: %x\n", pmic_id);
+
+		if (pmic_id != CRYSTALCOVE_PMIC) {
+			vibra_pwm_configure(info, false);
+		}
+	}
 
 	if (data->use_gpio_en) {
 		if (data->gpio_en < 0) {
-- 
1.9.1

